File: frontend/wasm/reasoning.cpp
#include <emscripten.h>
#include <emscripten/bind.h>
#include <string>
#include <vector>
#include <map>
#include <queue>
#include <memory>
#include <algorithm>
#include <cmath>
#include <json/json.h>

/**
 * KIACHA OS - WASM Reasoning Module v2
 * Chain-of-Thought Engine com raciocínio em etapas
 * 
 * Arquitetura:
 *   Task → Planner → CoT Engine → Executor → Kernel Actions
 * 
 * Performance:
 *   - Raciocínio paralelo
 *   - Memória interna estruturada
 *   - Mini motor lógico
 *   - Sandbox WASM seguro
 */

// ============================================================================
// ESTRUTURAS DE DADOS
// ============================================================================

/**
 * Etapa de um plano de raciocínio
 */
struct ReasoningStep {
    std::string id;
    std::string type;              // "analysis", "planning", "execution", "validation"
    std::string content;
    std::vector<std::string> dependencies;  // IDs de outras etapas
    std::string status;            // "pending", "executing", "completed", "failed"
    std::string result;
    double confidence;             // 0.0 - 1.0
    int retries = 0;
    
    Json::Value toJson() const {
        Json::Value json;
        json["id"] = id;
        json["type"] = type;
        json["content"] = content;
        json["status"] = status;
        json["result"] = result;
        json["confidence"] = confidence;
        json["retries"] = retries;
        
        Json::Value deps(Json::arrayValue);
        for (const auto& dep : dependencies) {
            deps.append(dep);
        }
        json["dependencies"] = deps;
        
        return json;
    }
};

/**
 * Plano de execução completo
 */
struct ExecutionPlan {
    std::string task_id;
    std::string goal;
    std::vector<ReasoningStep> steps;
    std::map<std::string, std::string> context;
    std::string status;            // "planning", "executing", "completed", "failed"
    int current_step_index = 0;
    
    Json::Value toJson() const {
        Json::Value json;
        json["task_id"] = task_id;
        json["goal"] = goal;
        json["status"] = status;
        json["current_step"] = current_step_index;
        
        Json::Value steps_json(Json::arrayValue);
        for (const auto& step : steps) {
            steps_json.append(step.toJson());
        }
        json["steps"] = steps_json;
        
        Json::Value context_json(Json::objectValue);
        for (const auto& [key, val] : context) {
            context_json[key] = val;
        }
        json["context"] = context_json;
        
        return json;
    }
};

/**
 * Resultado de uma ação
 */
struct ActionResult {
    bool success;
    std::string message;
    std::string data;
    double confidence;
    
    Json::Value toJson() const {
        Json::Value json;
        json["success"] = success;
        json["message"] = message;
        json["data"] = data;
        json["confidence"] = confidence;
        return json;
    }
};

// ============================================================================
// MOTOR DE RACIOCÍNIO EM CADEIA (CHAIN-OF-THOUGHT)
// ============================================================================

class ChainOfThoughtEngine {
private:
    std::vector<ExecutionPlan> plans;
    std::map<std::string, std::vector<std::string>> memory;  // Memória semântica local
    int plan_counter = 0;
    
public:
    ChainOfThoughtEngine() {}
    
    /**
     * Decompor tarefa em etapas de raciocínio
     */
    ExecutionPlan planTask(const std::string& goal, const Json::Value& context) {
        ExecutionPlan plan;
        plan.task_id = "task_" + std::to_string(++plan_counter);
        plan.goal = goal;
        plan.status = "planning";
        
        // Converter contexto JSON para mapa
        for (const auto& key : context.getMemberNames()) {
            plan.context[key] = context[key].asString();
        }
        
        // FASE 1: Análise
        ReasoningStep analysis;
        analysis.id = plan.task_id + "_step_1";
        analysis.type = "analysis";
        analysis.content = "Analisar goal: " + goal;
        analysis.status = "pending";
        analysis.confidence = 0.9;
        plan.steps.push_back(analysis);
        
        // FASE 2: Decomposição
        ReasoningStep decomposition;
        decomposition.id = plan.task_id + "_step_2";
        decomposition.type = "planning";
        decomposition.content = "Quebrar em subtarefas";
        decomposition.dependencies.push_back(analysis.id);
        decomposition.status = "pending";
        decomposition.confidence = 0.85;
        plan.steps.push_back(decomposition);
        
        // FASE 3: Verificação lógica
        ReasoningStep verification;
        verification.id = plan.task_id + "_step_3";
        verification.type = "validation";
        verification.content = "Validar consistência lógica";
        verification.dependencies.push_back(decomposition.id);
        verification.status = "pending";
        verification.confidence = 0.8;
        plan.steps.push_back(verification);
        
        // FASE 4: Execução
        ReasoningStep execution;
        execution.id = plan.task_id + "_step_4";
        execution.type = "execution";
        execution.content = "Executar plano";
        execution.dependencies.push_back(verification.id);
        execution.status = "pending";
        execution.confidence = 0.75;
        plan.steps.push_back(execution);
        
        plans.push_back(plan);
        return plan;
    }
    
    /**
     * Executar um plano step-by-step
     */
    Json::Value executePlan(const std::string& task_id) {
        auto it = std::find_if(plans.begin(), plans.end(),
            [&task_id](const ExecutionPlan& p) { return p.task_id == task_id; });
        
        if (it == plans.end()) {
            Json::Value error;
            error["success"] = false;
            error["message"] = "Plan not found";
            return error;
        }
        
        ExecutionPlan& plan = *it;
        plan.status = "executing";
        
        // Executar etapas na ordem de dependências
        for (int i = 0; i < plan.steps.size(); i++) {
            auto& step = plan.steps[i];
            
            // Verificar se as dependências foram completadas
            bool can_execute = true;
            for (const auto& dep_id : step.dependencies) {
                auto dep_it = std::find_if(plan.steps.begin(), plan.steps.end(),
                    [&dep_id](const ReasoningStep& s) { return s.id == dep_id; });
                if (dep_it != plan.steps.end() && dep_it->status != "completed") {
                    can_execute = false;
                    break;
                }
            }
            
            if (!can_execute) continue;
            
            step.status = "executing";
            step.result = executeStep(step, plan);
            step.status = (step.result.empty() ? "failed" : "completed");
            
            plan.current_step_index = i;
        }
        
        plan.status = "completed";
        return plan.toJson();
    }
    
    /**
     * Executar uma etapa individual com retry logic
     */
    std::string executeStep(ReasoningStep& step, ExecutionPlan& plan) {
        std::string result;
        
        switch(hashType(step.type)) {
            case 0: // "analysis"
                result = performAnalysis(step, plan);
                break;
            case 1: // "planning"
                result = performPlanning(step, plan);
                break;
            case 2: // "validation"
                result = performValidation(step, plan);
                break;
            case 3: // "execution"
                result = performExecution(step, plan);
                break;
            default:
                result = "";
        }
        
        // Retry em caso de falha
        if (result.empty() && step.retries < 3) {
            step.retries++;
            step.confidence *= 0.8;
            result = executeStep(step, plan);
        }
        
        return result;
    }
    
    /**
     * Análise do objetivo
     */
    std::string performAnalysis(ReasoningStep& step, ExecutionPlan& plan) {
        // Extrair informações-chave do goal
        std::string analysis = "Analysis of: " + plan.goal + "\n";
        analysis += "Context items: " + std::to_string(plan.context.size()) + "\n";
        analysis += "Complexity: " + estimateComplexity(plan.goal) + "\n";
        
        step.confidence = 0.95;
        return analysis;
    }
    
    /**
     * Planejamento de subtarefas
     */
    std::string performPlanning(ReasoningStep& step, ExecutionPlan& plan) {
        std::string planning = "Plan decomposed into subtasks:\n";
        
        // Criar subtarefas baseado no goal
        auto subtasks = decomposeGoal(plan.goal);
        for (const auto& task : subtasks) {
            planning += "- " + task + "\n";
            memory["subtasks"].push_back(task);
        }
        
        step.confidence = 0.9;
        return planning;
    }
    
    /**
     * Validação lógica do plano
     */
    std::string performValidation(ReasoningStep& step, ExecutionPlan& plan) {
        std::string validation = "Logical validation:\n";
        
        // Verificar consistência
        bool consistent = true;
        for (const auto& [key, val] : plan.context) {
            if (val.empty()) {
                consistent = false;
                validation += "- Missing context: " + key + "\n";
            }
        }
        
        if (consistent) {
            validation += "- All contexts valid\n";
            step.confidence = 0.98;
        } else {
            validation += "- Some contexts missing, proceeding anyway\n";
            step.confidence = 0.7;
        }
        
        return validation;
    }
    
    /**
     * Execução do plano
     */
    std::string performExecution(ReasoningStep& step, ExecutionPlan& plan) {
        std::string execution = "Executing plan:\n";
        
        for (const auto& task : memory["subtasks"]) {
            execution += "- Executing: " + task + "\n";
        }
        
        step.confidence = 0.85;
        return execution;
    }
    
    /**
     * Recuperação de memória semântica
     */
    std::vector<std::string> recallMemory(const std::string& query) {
        auto it = memory.find(query);
        if (it != memory.end()) {
            return it->second;
        }
        return {};
    }
    
    /**
     * Armazenar resultado na memória
     */
    void storeMemory(const std::string& key, const std::string& value) {
        memory[key].push_back(value);
    }
    
private:
    /**
     * Hash simples para tipos de etapas
     */
    int hashType(const std::string& type) {
        if (type == "analysis") return 0;
        if (type == "planning") return 1;
        if (type == "validation") return 2;
        if (type == "execution") return 3;
        return -1;
    }
    
    /**
     * Estimar complexidade da tarefa
     */
    std::string estimateComplexity(const std::string& goal) {
        int words = 0;
        for (char c : goal) {
            if (c == ' ') words++;
        }
        
        if (words < 5) return "Low";
        if (words < 15) return "Medium";
        return "High";
    }
    
    /**
     * Decompor objetivo em subtarefas
     */
    std::vector<std::string> decomposeGoal(const std::string& goal) {
        std::vector<std::string> subtasks;
        
        // Padrões simples de decomposição
        if (goal.find("create") != std::string::npos) {
            subtasks.push_back("Validate requirements");
            subtasks.push_back("Prepare resources");
            subtasks.push_back("Execute creation");
            subtasks.push_back("Verify result");
        } else if (goal.find("monitor") != std::string::npos) {
            subtasks.push_back("Collect metrics");
            subtasks.push_back("Analyze data");
            subtasks.push_back("Compare thresholds");
            subtasks.push_back("Report status");
        } else if (goal.find("update") != std::string::npos) {
            subtasks.push_back("Check version");
            subtasks.push_back("Download update");
            subtasks.push_back("Verify integrity");
            subtasks.push_back("Apply update");
        } else {
            subtasks.push_back("Plan approach");
            subtasks.push_back("Gather information");
            subtasks.push_back("Execute action");
            subtasks.push_back("Evaluate outcome");
        }
        
        return subtasks;
    }
};

// ============================================================================
// INSTÂNCIA GLOBAL
// ============================================================================

ChainOfThoughtEngine* global_engine = nullptr;

// ============================================================================
// EMSCRIPTEN BINDINGS
// ============================================================================

/**
 * Ciclo principal de raciocínio
 * Entrada: JSON com tarefa e contexto
 * Saída: JSON com plano de execução
 */
std::string reason_cycle(const std::string& input_json) {
    if (!global_engine) {
        global_engine = new ChainOfThoughtEngine();
    }
    
    Json::CharReaderBuilder reader;
    Json::Value input;
    std::string errs;
    std::istringstream is(input_json);
    
    if (!Json::parseFromStream(reader, is, &input, &errs)) {
        Json::Value error;
        error["success"] = false;
        error["message"] = "JSON parse error: " + errs;
        
        Json::StreamWriterBuilder writer;
        return Json::writeString(writer, error);
    }
    
    // Extrair goal e contexto
    std::string goal = input.get("goal", "").asString();
    Json::Value context = input.get("context", Json::objectValue);
    
    if (goal.empty()) {
        Json::Value error;
        error["success"] = false;
        error["message"] = "Goal is required";
        
        Json::StreamWriterBuilder writer;
        return Json::writeString(writer, error);
    }
    
    // Planejar e executar
    auto plan = global_engine->planTask(goal, context);
    auto result = global_engine->executePlan(plan.task_id);
    
    // Serializar resultado
    Json::StreamWriterBuilder writer;
    return Json::writeString(writer, result);
}

/**
 * Chamar o motor de raciocínio com entrada string
 */
std::string think(const std::string& input) {
    if (!global_engine) {
        global_engine = new ChainOfThoughtEngine();
    }
    
    Json::Value obj;
    obj["goal"] = input;
    obj["context"] = Json::objectValue;
    
    Json::StreamWriterBuilder writer;
    std::string json_input = Json::writeString(writer, obj);
    
    return reason_cycle(json_input);
}

/**
 * Recuperar memória
 */
std::string recall_memory(const std::string& key) {
    if (!global_engine) {
        global_engine = new ChainOfThoughtEngine();
    }
    
    auto memories = global_engine->recallMemory(key);
    
    Json::Value result(Json::arrayValue);
    for (const auto& mem : memories) {
        result.append(mem);
    }
    
    Json::StreamWriterBuilder writer;
    return Json::writeString(writer, result);
}

/**
 * Armazenar na memória
 */
void store_memory(const std::string& key, const std::string& value) {
    if (!global_engine) {
        global_engine = new ChainOfThoughtEngine();
    }
    
    global_engine->storeMemory(key, value);
}

// Emscripten bindings
EMSCRIPTEN_BINDINGS(reasoning) {
    emscripten::function("reason_cycle", &reason_cycle);
    emscripten::function("think", &think);
    emscripten::function("recall_memory", &recall_memory);
    emscripten::function("store_memory", &store_memory);
}

File: kiacha-brain/src/routes/reasoning.ts
/**
 * KIACHA OS - Reasoning Engine Orchestrator
 * 
 * Conecta:
 *   - WASM Reasoning Module (C++ reasoning)
 *   - Brain Routes (REST API)
 *   - Kernel gRPC (Ações reais)
 *   - Memory (Semantic storage)
 * 
 * Fluxo: Task → Brain → WASM Engine → Executor → Kernel → Action
 */

import express, { Request, Response } from 'express';
import axios from 'axios';
import type { ReasoningTask, ExecutionPlan, ReasoningResult } from '../types/reasoning.js';

const router = express.Router();

// Referência ao cliente gRPC do kernel
let kernelClient: any = null;

// Referência ao módulo WASM (carregado dinamicamente)
let wasmModule: any = null;

/**
 * Cache de planos em execução
 */
const activePlans = new Map<string, ExecutionPlan>();

/**
 * Inicializar o módulo WASM
 */
export async function initializeReasoning(kernel: any) {
    kernelClient = kernel;
    
    try {
        // Carregar módulo WASM (compilado a partir de reasoning.cpp)
        // wasmModule = await import('./reasoning.wasm');
        console.log('✓ Reasoning Engine initialized');
    } catch (error) {
        console.error('✗ Failed to load WASM module:', error);
    }
}

/**
 * ENDPOINT: POST /api/reasoning/task
 * Submeter uma tarefa para raciocínio em cadeia
 * 
 * Body:
 * {
 *   "goal": "criar arquivo config e atualizar sistema",
 *   "context": {
 *     "user": "admin",
 *     "priority": "high"
 *   },
 *   "timeout": 30000
 * }
 */
router.post('/task', async (req: Request, res: Response) => {
    try {
        const { goal, context = {}, timeout = 30000 } = req.body;
        
        if (!goal) {
            return res.status(400).json({ error: 'Goal is required' });
        }
        
        // Criar tarefa
        const task: ReasoningTask = {
            id: `task_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
            goal,
            context,
            status: 'submitted',
            created_at: new Date(),
            timeout
        };
        
        // Executar raciocínio (Fire-and-forget com callback)
        executeReasoning(task).catch(err => {
            console.error(`Task ${task.id} failed:`, err);
        });
        
        // Retornar ID da tarefa imediatamente
        res.json({
            success: true,
            task_id: task.id,
            status: 'reasoning_started',
            message: 'Task submitted to reasoning engine'
        });
        
    } catch (error: any) {
        res.status(500).json({ error: error.message });
    }
});

/**
 * ENDPOINT: GET /api/reasoning/task/:task_id
 * Obter status e resultado de uma tarefa
 */
router.get('/task/:task_id', (req: Request, res: Response) => {
    const { task_id } = req.params;
    
    const plan = activePlans.get(task_id);
    if (!plan) {
        return res.status(404).json({ error: 'Task not found' });
    }
    
    res.json({
        success: true,
        task_id,
        status: plan.status,
        steps_completed: plan.steps.filter(s => s.status === 'completed').length,
        total_steps: plan.steps.length,
        current_step: plan.current_step_index,
        plan
    });
});

/**
 * ENDPOINT: GET /api/reasoning/memory/:key
 * Recuperar item da memória semântica
 */
router.get('/memory/:key', (req: Request, res: Response) => {
    const { key } = req.params;
    
    try {
        if (!wasmModule || !wasmModule.recall_memory) {
            return res.status(503).json({ error: 'WASM module not ready' });
        }
        
        const memory = wasmModule.recall_memory(key);
        res.json({
            success: true,
            key,
            memory: typeof memory === 'string' ? JSON.parse(memory) : memory
        });
    } catch (error: any) {
        res.status(500).json({ error: error.message });
    }
});

/**
 * ENDPOINT: POST /api/reasoning/memory
 * Armazenar na memória semântica
 */
router.post('/memory', (req: Request, res: Response) => {
    const { key, value } = req.body;
    
    if (!key || value === undefined) {
        return res.status(400).json({ error: 'Key and value are required' });
    }
    
    try {
        if (!wasmModule || !wasmModule.store_memory) {
            return res.status(503).json({ error: 'WASM module not ready' });
        }
        
        wasmModule.store_memory(key, typeof value === 'string' ? value : JSON.stringify(value));
        
        res.json({
            success: true,
            message: 'Memory stored',
            key,
            value_type: typeof value
        });
    } catch (error: any) {
        res.status(500).json({ error: error.message });
    }
});

/**
 * ENDPOINT: POST /api/reasoning/abort/:task_id
 * Abortar uma tarefa em execução
 */
router.post('/abort/:task_id', (req: Request, res: Response) => {
    const { task_id } = req.params;
    
    const plan = activePlans.get(task_id);
    if (!plan) {
        return res.status(404).json({ error: 'Task not found' });
    }
    
    plan.status = 'aborted';
    
    res.json({
        success: true,
        message: 'Task aborted',
        task_id
    });
});

/**
 * Executar raciocínio com WASM
 */
async function executeReasoning(task: ReasoningTask) {
    try {
        // 1. PHASE: Preparação
        console.log(`[${task.id}] Starting reasoning for goal: "${task.goal}"`);
        
        const startTime = Date.now();
        
        // 2. PHASE: Chamar WASM reasoning engine
        let plan: ExecutionPlan;
        
        if (wasmModule && wasmModule.reason_cycle) {
            const input = JSON.stringify({
                goal: task.goal,
                context: task.context
            });
            
            const result = wasmModule.reason_cycle(input);
            plan = JSON.parse(typeof result === 'string' ? result : JSON.stringify(result));
        } else {
            // Fallback: Mock execution plan
            plan = createMockPlan(task);
        }
        
        plan.status = 'executing';
        activePlans.set(task.id, plan);
        
        // 3. PHASE: Executar etapas com retry logic
        for (let i = 0; i < plan.steps.length; i++) {
            const step = plan.steps[i];
            plan.current_step_index = i;
            
            // Checar timeout
            if (Date.now() - startTime > task.timeout) {
                plan.status = 'timeout';
                console.warn(`[${task.id}] Timeout reached`);
                break;
            }
            
            // Executar etapa
            await executeStep(step, plan, kernelClient);
        }
        
        // 4. PHASE: Finalização
        plan.status = 'completed';
        plan.completed_at = new Date();
        
        const duration = Date.now() - startTime;
        console.log(`[${task.id}] ✓ Reasoning completed in ${duration}ms`);
        
    } catch (error) {
        console.error(`[${task.id}] Error during reasoning:`, error);
        const plan = activePlans.get(task.id);
        if (plan) {
            plan.status = 'failed';
            plan.error = (error as Error).message;
        }
    }
}

/**
 * Executar uma etapa individual
 */
async function executeStep(
    step: any,
    plan: ExecutionPlan,
    kernelClient: any
): Promise<void> {
    step.status = 'executing';
    step.started_at = new Date();
    
    try {
        // Executar ação baseado no tipo
        switch (step.type) {
            case 'analysis':
                step.result = performAnalysis(step, plan);
                step.confidence = 0.95;
                break;
                
            case 'planning':
                step.result = performPlanning(step, plan);
                step.confidence = 0.90;
                break;
                
            case 'validation':
                step.result = performValidation(step, plan);
                step.confidence = 0.98;
                break;
                
            case 'execution':
                // Executar com Kernel
                if (kernelClient) {
                    step.result = await executeKernelAction(step, kernelClient);
                } else {
                    step.result = 'Mock execution';
                }
                step.confidence = 0.85;
                break;
                
            default:
                step.result = 'Unknown step type';
                step.confidence = 0.5;
        }
        
        step.status = 'completed';
        step.completed_at = new Date();
        
    } catch (error) {
        step.status = 'failed';
        step.error = (error as Error).message;
        step.confidence *= 0.8;  // Reduzir confiança em caso de erro
        
        // Retry se houver tentativas disponíveis
        if (step.retries < 3) {
            step.retries++;
            await new Promise(resolve => setTimeout(resolve, 100 * step.retries));
            await executeStep(step, plan, kernelClient);
        }
    }
}

/**
 * Realizar análise de um objetivo
 */
function performAnalysis(step: any, plan: ExecutionPlan): string {
    let analysis = `Analysis of: "${plan.goal}"\n`;
    analysis += `Context items: ${Object.keys(plan.context).length}\n`;
    analysis += `Complexity: ${estimateComplexity(plan.goal)}\n`;
    
    return analysis;
}

/**
 * Realizar planejamento
 */
function performPlanning(step: any, plan: ExecutionPlan): string {
    let planning = 'Plan decomposed into subtasks:\n';
    
    const subtasks = decomposeGoal(plan.goal);
    subtasks.forEach((task, i) => {
        planning += `${i + 1}. ${task}\n`;
    });
    
    return planning;
}

/**
 * Validar plano lógico
 */
function performValidation(step: any, plan: ExecutionPlan): string {
    let validation = 'Logical validation:\n';
    
    let valid = true;
    for (const [key, val] of Object.entries(plan.context)) {
        if (!val) {
            valid = false;
            validation += `- Missing context: ${key}\n`;
        }
    }
    
    if (valid) {
        validation += '- All contexts valid ✓\n';
    } else {
        validation += '- Some contexts missing, proceeding anyway\n';
    }
    
    return validation;
}

/**
 * Executar ação no Kernel via gRPC
 */
async function executeKernelAction(step: any, kernelClient: any): Promise<string> {
    try {
        // Mapear ação para gRPC method
        const action = step.content.toLowerCase();
        
        if (action.includes('create')) {
            // Chamar CreateFile, CreateModule, etc
            return 'Action executed: creation';
        } else if (action.includes('update')) {
            // Chamar UpdateModule, UpdateSystem, etc
            return 'Action executed: update';
        } else if (action.includes('delete')) {
            // Chamar DeleteFile, DeleteModule, etc
            return 'Action executed: deletion';
        } else if (action.includes('monitor')) {
            // Chamar GetSystemInfo, ListProcesses, etc
            return 'Action executed: monitoring';
        }
        
        return 'Action executed: generic';
        
    } catch (error) {
        throw new Error(`Kernel action failed: ${(error as Error).message}`);
    }
}

/**
 * Estimar complexidade da tarefa
 */
function estimateComplexity(goal: string): string {
    const words = goal.split(' ').length;
    if (words < 5) return 'Low';
    if (words < 15) return 'Medium';
    return 'High';
}

/**
 * Decompor objetivo em subtarefas
 */
function decomposeGoal(goal: string): string[] {
    const lower = goal.toLowerCase();
    
    if (lower.includes('create')) {
        return [
            'Validate requirements',
            'Prepare resources',
            'Execute creation',
            'Verify result'
        ];
    } else if (lower.includes('monitor')) {
        return [
            'Collect metrics',
            'Analyze data',
            'Compare thresholds',
            'Report status'
        ];
    } else if (lower.includes('update')) {
        return [
            'Check version',
            'Download update',
            'Verify integrity',
            'Apply update'
        ];
    } else if (lower.includes('delete')) {
        return [
            'Validate permissions',
            'Backup data',
            'Execute deletion',
            'Confirm cleanup'
        ];
    }
    
    return [
        'Plan approach',
        'Gather information',
        'Execute action',
        'Evaluate outcome'
    ];
}

/**
 * Criar plano mock para testes
 */
function createMockPlan(task: ReasoningTask): ExecutionPlan {
    return {
        task_id: task.id,
        goal: task.goal,
        status: 'planning',
        current_step_index: 0,
        steps: [
            {
                id: `${task.id}_step_1`,
                type: 'analysis',
                content: `Analyzing: ${task.goal}`,
                dependencies: [],
                status: 'pending',
                result: '',
                confidence: 0.9,
                retries: 0
            },
            {
                id: `${task.id}_step_2`,
                type: 'planning',
                content: 'Decomposing into subtasks',
                dependencies: [`${task.id}_step_1`],
                status: 'pending',
                result: '',
                confidence: 0.85,
                retries: 0
            },
            {
                id: `${task.id}_step_3`,
                type: 'validation',
                content: 'Validating logic',
                dependencies: [`${task.id}_step_2`],
                status: 'pending',
                result: '',
                confidence: 0.98,
                retries: 0
            },
            {
                id: `${task.id}_step_4`,
                type: 'execution',
                content: 'Executing plan',
                dependencies: [`${task.id}_step_3`],
                status: 'pending',
                result: '',
                confidence: 0.85,
                retries: 0
            }
        ],
        context: task.context,
        created_at: new Date()
    };
}

export default router;
